File: bot.py
"""
OpenClaw Bot - Main Entry Point

A bot that assists with planning on Notion and managing tasks on GitHub repositories.
"""
import sys
from config import Config
from notion_integration import NotionIntegration
from github_integration import GitHubIntegration


class OpenClawBot:
    """Main bot class that orchestrates Notion and GitHub integrations"""
    
    def __init__(self):
        """Initialize the bot with required integrations"""
        print(f"Initializing {Config.BOT_NAME}...")
        
        # Validate configuration
        missing_config = Config.validate()
        if missing_config:
            print(f"Error: Missing required configuration: {', '.join(missing_config)}")
            print("Please create a .env file based on .env.example and fill in the required values.")
            sys.exit(1)
        
        # Initialize integrations
        try:
            self.notion = NotionIntegration()
            self.github = GitHubIntegration()
            print("✓ Bot initialized successfully!")
        except Exception as e:
            print(f"Error initializing bot: {e}")
            sys.exit(1)
    
    def _extract_task_title(self, task):
        """
        Extract task title from Notion task object
        
        Args:
            task (dict): Notion task object
            
        Returns:
            str: Task title or 'Untitled' if not found
        """
        try:
            title_data = task.get('properties', {}).get('Name', {}).get('title', [])
            if title_data:
                return title_data[0].get('text', {}).get('content', 'Untitled')
        except (KeyError, IndexError, TypeError):
            pass
        return 'Untitled'
    
    def sync_notion_to_github(self):
        """
        Sync tasks from Notion to GitHub issues
        
        This method retrieves tasks from Notion and creates corresponding
        GitHub issues for tracking and assignment.
        """
        print("\n--- Syncing Notion tasks to GitHub ---")
        
        # Get tasks from Notion
        notion_tasks = self.notion.get_tasks()
        print(f"Found {len(notion_tasks)} tasks in Notion")
        
        # Process each task
        for task in notion_tasks:
            title = self._extract_task_title(task)
            print(f"  - {title}")
        
        return notion_tasks
    
    def sync_github_to_notion(self):
        """
        Sync GitHub issues back to Notion
        
        This method retrieves GitHub issues and updates corresponding
        Notion tasks with the latest status.
        """
        print("\n--- Syncing GitHub issues to Notion ---")
        
        # Get issues from GitHub
        github_issues = self.github.get_issues()
        print(f"Found {len(github_issues)} open issues in GitHub")
        
        # Process each issue
        for issue in github_issues:
            print(f"  - #{issue.number}: {issue.title}")
        
        return github_issues
    
    def create_task_from_github_issue(self, issue_number):
        """
        Create a Notion task from a GitHub issue
        
        Args:
            issue_number (int): GitHub issue number
        """
        print(f"\n--- Creating Notion task from GitHub issue #{issue_number} ---")
        
        # Get the issue
        issue = self.github.repo.get_issue(issue_number)
        
        # Create corresponding Notion task
        task = self.notion.create_task(
            title=issue.title,
            description=issue.body or "No description provided",
            status="In Progress" if issue.state == "open" else "Done"
        )
        
        if task:
            print(f"✓ Created Notion task for issue #{issue_number}")
        else:
            print(f"✗ Failed to create Notion task")
        
        return task
    
    def create_github_issue_from_task(self, task_title, task_description=None):
        """
        Create a GitHub issue from a Notion task
        
        Args:
            task_title (str): Task title
            task_description (str): Task description
        """
        print(f"\n--- Creating GitHub issue from task: {task_title} ---")
        
        issue = self.github.create_issue(
            title=task_title,
            body=task_description
        )
        
        if issue:
            print(f"✓ Created GitHub issue #{issue.number}")
        else:
            print(f"✗ Failed to create GitHub issue")
        
        return issue
    
    def run(self):
        """
        Main bot execution loop
        
        This is a placeholder for the main bot logic.
        You can customize this to run scheduled tasks, listen to webhooks, etc.
        """
        print(f"\n{Config.BOT_NAME} is running!")
        print("\nAvailable operations:")
        print("  1. Sync Notion tasks to GitHub")
        print("  2. Sync GitHub issues to Notion")
        print("  3. Create GitHub issue from Notion task")
        print("  4. Create Notion task from GitHub issue")
        print("\nFor automated syncing, you can schedule these operations using cron or a task scheduler.")


def main():
    """Main entry point"""
    try:
        bot = OpenClawBot()
        bot.run()
        
        # Example: Show current tasks and issues
        bot.sync_notion_to_github()
        bot.sync_github_to_notion()
        
    except KeyboardInterrupt:
        print("\n\nBot stopped by user.")
    except Exception as e:
        print(f"\nError running bot: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/bot.rs
/// OpenClaw Bot - Main Bot Logic
use anyhow::Result;
use crate::config::Config;
use crate::github::GitHubIntegration;
use crate::notion::NotionIntegration;
use octocrab::params::State;

pub struct OpenClawBot {
    config: Config,
    notion: NotionIntegration,
    github: GitHubIntegration,
}

impl OpenClawBot {
    /// Initialize the bot with required integrations
    pub fn new(config: Config) -> Result<Self> {
        println!("Initializing {}...", config.bot_name);
        
        // Initialize integrations
        let notion = NotionIntegration::new(
            config.notion_api_key.clone(),
            config.notion_database_id.clone(),
        );
        
        let github = GitHubIntegration::new(
            config.github_token.clone(),
            config.github_owner.clone(),
            config.github_repo.clone(),
        )?;
        
        println!("✓ Bot initialized successfully!");
        
        Ok(Self {
            config,
            notion,
            github,
        })
    }
    
    /// Extract task title from Notion task object
    fn extract_task_title(&self, task: &crate::notion::NotionTask) -> String {
        task.properties
            .get("Name")
            .and_then(|name| name.get("title"))
            .and_then(|title| title.as_array())
            .and_then(|arr| arr.first())
            .and_then(|first| first.get("text"))
            .and_then(|text| text.get("content"))
            .and_then(|content| content.as_str())
            .unwrap_or("Untitled")
            .to_string()
    }
    
    /// Sync tasks from Notion to GitHub issues
    pub fn sync_notion_to_github(&self) -> Result<Vec<crate::notion::NotionTask>> {
        println!("\n--- Syncing Notion tasks to GitHub ---");
        
        // Get tasks from Notion
        let notion_tasks = self.notion.get_tasks()?;
        println!("Found {} tasks in Notion", notion_tasks.len());
        
        // Process each task
        for task in &notion_tasks {
            let title = self.extract_task_title(task);
            println!("  - {}", title);
        }
        
        Ok(notion_tasks)
    }
    
    /// Sync GitHub issues back to Notion
    pub async fn sync_github_to_notion(&self) -> Result<Vec<octocrab::models::issues::Issue>> {
        println!("\n--- Syncing GitHub issues to Notion ---");
        
        // Get issues from GitHub
        let github_issues = self.github.get_issues(Some(State::Open)).await?;
        println!("Found {} open issues in GitHub", github_issues.len());
        
        // Process each issue
        for issue in &github_issues {
            println!("  - #{}: {}", issue.number, issue.title);
        }
        
        Ok(github_issues)
    }
    
    /// Create a Notion task from a GitHub issue
    pub async fn create_task_from_github_issue(&self, issue_number: u64) -> Result<crate::notion::NotionTask> {
        println!("\n--- Creating Notion task from GitHub issue #{} ---", issue_number);
        
        // Get the issue
        let issues = self.github.get_issues(None).await?;
        let issue = issues
            .iter()
            .find(|i| i.number == issue_number)
            .ok_or_else(|| anyhow::anyhow!("Issue #{} not found", issue_number))?;
        
        // Create corresponding Notion task
        let status = match issue.state {
            octocrab::models::IssueState::Open => "In Progress",
            octocrab::models::IssueState::Closed => "Done",
            _ => "Not Started",
        };
        
        let description = issue.body.as_deref();
        let task = self.notion.create_task(&issue.title, description, status)?;
        
        println!("✓ Created Notion task for issue #{}", issue_number);
        
        Ok(task)
    }
    
    /// Create a GitHub issue from a Notion task
    pub async fn create_github_issue_from_task(&self, task_title: &str, task_description: Option<&str>) -> Result<octocrab::models::issues::Issue> {
        println!("\n--- Creating GitHub issue from task: {} ---", task_title);
        
        let issue = self.github.create_issue(task_title, task_description, None, None).await?;
        
        println!("✓ Created GitHub issue #{}", issue.number);
        
        Ok(issue)
    }
    
    /// Main bot execution loop
    pub async fn run(&self) -> Result<()> {
        println!("\n{} is running!", self.config.bot_name);
        println!("\nAvailable operations:");
        println!("  1. Sync Notion tasks to GitHub");
        println!("  2. Sync GitHub issues to Notion");
        println!("  3. Create GitHub issue from Notion task");
        println!("  4. Create Notion task from GitHub issue");
        println!("\nFor automated syncing, you can schedule these operations using cron or a task scheduler.");
        println!("\n--- Running sync operations ---");
        
        // Example: Show current tasks and issues
        if let Err(e) = self.sync_notion_to_github() {
            eprintln!("Error syncing Notion to GitHub: {}", e);
        }
        
        if let Err(e) = self.sync_github_to_notion().await {
            eprintln!("Error syncing GitHub to Notion: {}", e);
        }
        
        Ok(())
    }
}
